### PR TITLE
[GCS] Use gRPC instead of socket for GCS client health check

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
A user has reported a crash in GCS client where the client was unable to connect to the GCS server after retries, even when GCS server has always been running. I was not able to reproduce the exact issue, but noticed that the health check logic with socket has unexpected behavior sometimes, e.g. it is much slower to use socket for health check compared to using gRPC (~40s vs < 1s sometimes). The user issue could be related to this slowness, so this PR updates the logic to use gRPC health check.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 1]
```diff
@@ -180,10 +180,38 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() {
   return current_gcs_server_address_;
 }
 
+/// Checks whether GCS at the specified address is healthy.
+bool GcsClient::CheckHealth(const std::string &ip, int port, int64_t timeout_ms) {
+  // Health checking currently needs to be blocking. So it cannot use the event loop
+  // in the GcsClient. Otherwise there can be deadlocks.
+  auto channel = grpc::CreateChannel(absl::StrCat(ip, ":", port),
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:187]: How expensive is this operation compared with previous ones?

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 2]
```diff
@@ -180,10 +180,38 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() {
...
+    return false;
+  }
+  return true;
+}
+
 void GcsClient::PeriodicallyCheckGcsServerAddress() {
   if (disconnected_) {
     return;
   }
+  if (current_connection_failure_.has_value()) {
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:210]: I feel this logic is a little bit hard to follow. This function seems to mix handling server addr change and connection error and this function is called periodically which doesn't seem very fit doing handling error change.

Maybe we should check the logic here and maybe come up with a better flow of doing the work.

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 1 above)
Comment (mwtian) [reply src/ray/gcs/gcs_client/gcs_client.cc:187]: This is faster than the socket approach when I tested them.

File: src/ray/gcs/gcs_client/gcs_client.cc (hunk 2 above)
Comment (mwtian) [reply src/ray/gcs/gcs_client/gcs_client.cc:211]: Renamed the function to `PeriodicallyCheckGcsConnection()`. The problem with the existing approach to handle GCS client rpc error is that if there is a connection issue, all inflight requests would fail and sequentially each response callback would try to reconnect with GCS. This would slow down the client and may trigger unexpected behavior with the socket connection test. It seems better to have only one callback responsible to trying to reconnect with GCS. Alternatively, we can create another periodic runner for this purpose. Or we can revert to the existing behavior.

**Comment by mwtian** [comment]
Later, we can look into making the reconnection logic async, and ensuring at most 1 inflight reconnection.

File: src/ray/gcs/gcs_client/gcs_client.cc [hunk 3]
```diff
@@ -180,15 +180,43 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() {
...
+  if (!status.ok()) {
+    RAY_LOG(WARNING) << "Unable to reach GCS at " << ip << ":" << port
+                     << ". Failure: " << status.error_code() << " "
+                     << status.error_message();
+    return false;
+  }
+  return true;
+}
+
+void GcsClient::PeriodicallyCheckGcsConnection() {
```
Comment (iycheng) [inline src/ray/gcs/gcs_client/gcs_client.cc:206]: Synced offline. I'm good about this change now since we'll refactor (cleanup) this part later during GCS HA. Could you add some todo here? I'll be good to let it go.

Comment (mwtian) [reply src/ray/gcs/gcs_client/gcs_client.cc:207]: Added.

**Review by fishbone — APPROVED** [review]
